cirq/circuits/sparse_circuit_dag.py

```diff
@@ -0,0 +1,247 @@
+        (2, cirq.CZ(cirq.LineQubit(0), cirq.LineQubit(1))),
+        (3, cirq.H(cirq.LineQubit(3))),
+        (4, cirq.H(cirq.LineQubit(4))),
+        and the folllowing sequence of frontiers:
+        (
+        (1, cirq.H(cirq.LineQubit(1))), (1, cirq.H(cirq.LineQubit(2)))
+        ),
+        (
+        (2, cirq.CZ(cirq.LineQubit(0), cirq.LineQubit(1))),
+        (1, cirq.H(cirq.LineQubit(2)))
+        ),
+        (
+        (3, cirq.H(cirq.LineQubit(0))),
+        (2, cirq.CZ(cirq.LineQubit(0), cirq.LineQubit(1))),
+        (1, cirq.H(cirq.LineQubit(2)))
+        )
+        The last iteration invoves two pairs:
+        (4, cirq.H(cirq.LineQubit(0))),
+        (4, cirq.CZ(cirq.LineQubit(1), cirq.LineQubit(2)))
+        of which the first is connected to the first two and the second
+        is connected to the second two nodes from the last frontier.
+        In this case, the circuit has a single independent factor
+        (i.e. it cannot be factorized)
+
+        """
+        dag = SparseCircuitDag(can_reorder=can_reorder, device=device)
+        cur_index = 0
+        # All qubits participating in the operations before the current moment.
+        all_qubits: Set['cirq.Qid'] = set()
+        # Frontier is keeping track of most recent operations on all qubits
+        # which participated in the operations before the current moment.
+        # These are the only "nodes" which could be connected to the operatioins
+        # present in the current moment.
+        frontier: Set[Tuple[int, 'cirq.Operation']] = set()
+        # Next iteration of the frontier.
+        next_frontier = set()
+        # Iterate over (op, moment_num) pairs
+        # (the assumption is that the moments are listed in order).
+        # Keep track of all qubits which are, directly or indirectly,
+        # affected by these operations.
+        for index, op in operations_by_moment_index:
+            # Assumpiton: next index corresponds to the next moment
+            # in the circuit
+            if index != cur_index:
+                # Expectation: the sequence is ordered by moment indices.
+                print("! index {}, cur_index {}".format(index, cur_index))
+                if index < cur_index:
+                    raise ValueError("Moment indices expected to increase.")
+
+                for prev_node in frontier:
+                    if any(q not in all_qubits for q in prev_node[1].qubits):
+                        next_frontier.add(prev_node)
+                frontier = next_frontier
+                next_frontier = set()
+                all_qubits = set()
+                cur_index = index
+            print("index {}, cur_index {}".format(index, cur_index))
+            new_node = (index, op)
+            dag.add_node(new_node)
+            next_frontier.add(new_node)
+            all_qubits |= set(op.qubits)
+            for prev_node in frontier:
+                if not dag.can_reorder(prev_node, new_node):
```

Comment:
This is quadratic on number of qubits. If the default implementation is only add edges for overlapping qubit sets, it should be possible to maintain a mapping from qubit to all the frontier ops that contain it, and run more efficiently. IDK if anyone would plan to simulate an algo with thousands of qubits though.

Reply:
Yeah, well noted. It's a small improvement with regard to CircuitDag, that loops over all nodes: https://github.com/quantumlib/Cirq/blob/a61e51b53612735e93b3bb8a7605030c499cd6c7/cirq/circuits/circuit_dag.py#L121 - but I agree that this can be made better.